### ISSUE TITLE
Jaccard and FiScore metrics aren't working properly in 0.8 version

### ISSUE THREAD

**Issue by PCPJ** [PyTorchLightning/metrics#977]
## 🐛 Bug

I am using the IoU and F1 metrics to evaluate my segmentation model. Everything was running fine, until the 0.8 version was released.
I just changed the Metric names from IoU to JaccardIndex, and F1 to F1Score. The problem is that my the metric calculation looks random now.

Before 0.8:
![before](https://user-images.githubusercontent.com/8596365/164522205-c66ac3e1-8792-4127-8955-17c58c630c80.png)
After 0.8:
![after](https://user-images.githubusercontent.com/8596365/164522254-34dcd19c-8b40-47cc-86c5-704531f4391c.png)

I don't know if I miss something in this new patch, but all the rest of my code is the same. The only change is the torchmetrics version and the metrics names.

### To Reproduce

Without sharing my dataset, I dont know how someone else could reproduce my problem. 
Maybe I can use a small public dataset with both version of torchmetric, if the discrepancy continues, someone else will be able to reproduce it.

#### Code sample

This is the classes I have been using. Only thing changed is the IoU to JaccardIndex and F1 to F1Score.
```python
class IOUMetric(torchmetrics.Metric):
    def __init__(self, num_classes):
        super().__init__(dist_sync_on_step=False, compute_on_step=False)

        self.iou = torchmetrics.JaccardIndex(num_classes=num_classes,
                                     reduction="elementwise_mean",
                                     compute_on_step=False,
                                     dist_sync_on_step=False)

    def update(self, preds: torch.Tensor, targets: torch.Tensor):
        return self.iou(preds, targets)

    def compute(self):
        return self.iou.compute().item()

    def reset(self):
        self.iou.reset()

class F1Metric(torchmetrics.Metric):
    def __init__(self, num_classes):
        super().__init__(dist_sync_on_step=False, compute_on_step=False)

        self.f1 = torchmetrics.F1Score(num_classes=num_classes,
                                     average="micro",
                                     mdmc_average="samplewise",
                                     compute_on_step=False,
                                     dist_sync_on_step=False)
        
    def update(self, preds: torch.Tensor, targets: torch.Tensor):
        return self.f1(preds, targets)

    def compute(self):
        return self.f1.compute().item()

    def reset(self):
        self.f1.reset()
```
I am using them with a MatricCollection:
```python
def create_metrics(names, conf):
    metrics = list()
    for name in names:
        if name == "IOU":
            metrics.append(create_IOU(conf[name]))
        elif name == "F1":
            metrics.append(create_F1(conf[name]))
    return MetricCollection(metrics).to(device)
```

The tensors I am sending thought the metrics have shape:
pred => (batch, classes, rows, cols) of floats
targets => (batch, rows, cols) of integer

**Comment by SkafteNicki** [comment]
Hi @PCPJ, thanks for raising this issue.
At first glance I cannot reproduce the behaviour you are reporting. If I run the following script:
```python
import torchmetrics
import torch

if __name__ == "__main__":
    print(torchmetrics.__version__)
    _ = torch.manual_seed(42)

    num_classes = 5

    if "0.7" in torchmetrics.__version__:
        metric = torchmetrics.F1(
            num_classes=num_classes,
            average="micro",
            mdmc_average="samplewise",
        )
    else:
        metric = torchmetrics.F1Score(
            num_classes=num_classes,
            average="micro",
            mdmc_average="samplewise",
        )

    preds = torch.randn(100, 5, 28, 28)
    target = torch.randint(5, (100, 28, 28))

    val = metric(preds, target)
    print(val)

    if "0.7" in torchmetrics.__version__:
        metric = torchmetrics.IoU(
            num_classes=num_classes,
            reduction="elementwise_mean",
        )
    else:
        metric = torchmetrics.JaccardIndex(
            num_classes=num_classes,
            reduction="elementwise_mean",
        )
    val = metric(preds, target)
    print(val)
```
I get the same values regardless of version. 
Only thing I can think of is that `MetricCollection` received a pretty huge update in v0.8 which might have introduced some unwanted bugs. Could you please try to initialize them with the new compute group feature disabled:
```python
# Instead of
MetricCollection(metrics).to(device)
# do
MetricCollection(metrics, compute_groups=False).to(device)
```
and report back if your results look as they should.

**Comment by SkafteNicki** [comment]
Hi @PCPJ,
Is your code in a public repository, because then I can take a look at what the problem might be?

**Comment by PCPJ** [comment]
Thanks for your quick reply @SkafteNicki.
No, my code isn't in a public repository. What I did was taking my code, cropping all the unnecessary stuff and committing in a public repository.
[Here it is](https://github.com/PCPJ/Torchmetrics-BUG).
I tested this code, and it is presenting difference between versions. The graphs showing the difference is in the repo as well.
Thank you so much.

**Comment by SkafteNicki** [comment]
Hi PCPJ,
I finally got time to debug this and I have some good news and some bad news:

The metrics never stopped working. We have CI to test this so I was sceptical that they would begin to produce wrong results.
What has changed is that the argument `compute_on_step` has be deprecated/removed. If you take your code with v0.7 of torchmetrics and set `compute_on_step=True` you can basically reproduce the behaviour in v0.8. The reason that `compute_on_step` was deprecated was that `compute_on_step=False` is exactly the same as calling just `update`.

Or so I though. There was a difference when you are working with nested metrics, as in your example because of the way we are caching metric states. This made me realize that we DO NOT support nested metrics (we do not explicit state this anywhere), because our code would only cache the metric state of the top-level metric and not everything nested which can lead to wrong results when using `forward`.

You have two solutions for your code. Either exchange `forward` with `update`:
```python
metrics(preds, target)
# to
metrics.update(preds, target)
```
or instead of working with nested metrics, simply subclass instead and add whatever method you need:
```python
# this is the IOU example metric you been using but written as a subclass instead
class IOUMetric(torchmetrics.JaccardIndex):
    def __init__(self, num_classes):
        super().__init__(num_classes=num_classes, reduction="elementwise_mean")

    def to_string(self, results):
        return "{:.5f}".format(results)

    def save_graph(self, out_dir, model_name, epoch, train_records, val_records, test_records, test_sets_names):
        save_graph(os.path.join(out_dir, model_name+"_IOUMetric.png"), epoch, train_records, val_records, test_records, test_sets_names, label="IOUMetric")

    def save_if_best(self, cur, best, check_point, path):
        if best is None or cur >= best:
            path = path[:-4] + "_best-IOUMetric.pth"
            torch.save(check_point, path)
            return cur
        return best
```
@Borda and @justusschock what are you opinions on nested metrics? I feel like we already support this enough through `MetricCollection` and `MetricComposition` and we do not need to support stuff like:
```python
class MyMetric(torchmetrics.Metric):
    def __init__(self, num_classes)
        self.metric = torchmetrics.Accuracy()
        # some other stuff

    def update(self, *args, **kwargs):
        self.metric.update(*args, **kwargs)
```
If you agree, we probably need to be more explicit that we do not out of the box support nested metrics. We could also consider the possibility of directly checking this and raising a error/warning.

**Comment by justusschock** [comment]
@SkafteNicki I think so as well. Instead of nesting them, there is always the way of subclassing for extension. So I think we shouldn't  forbid it for single metrics but not put in effort to support this by collections
